MF-64 Express and verify backend dependencies for frontend modules

fatmali opened the pull request.

fatmali committed:
MF-64 Express and verify backend dependencies for frontend modules

fatmali commented on src/openmrs-backend-dependencies.ts:
```diff
@@ -0,0 +1,91 @@
+  const modulesWithWrongVersionInstalled = getMismatchedVersions(
+    requiredBackendModules
+  );
+  console.error(
```
Will replace these with displaying on the UI/ throwing the error

fatmali replied:
We will replace the console error messages by either throwing the error or displaying the errors on devtools ui

fatmali committed:
Self review

fatmali committed:
Fix travis build

internettrans commented on package.json:
```diff
@@ -56,16 +56,17 @@
     "kremling": "^2.0.1",
     "react": "^16.9.0",
     "react-dom": "^16.9.0",
+    "semver": "^6.3.0",
```
`semver` should be a `devDependency`.

`dependencies` are for "in-browser" dependencies. `devDependencies` are for everything else.

internettrans commented on src/declarations.d.tsx:
```diff
@@ -8,3 +8,5 @@ declare module "*.css" {
   export default styles;
 }
+
+declare var System: any;
```
Alternatively, you can `npm i -S @types/systemjs` I think

internettrans commented on src/openmrs-backend-dependencies.ts:
```diff
@@ -0,0 +1,92 @@
+System.constructor.prototype.onload = function(err, id, deps) {
+  if (!err) {
+    System.import(id).then(module => {
+      console.log(module);
```
This console.log shouldn't be here.

internettrans commented on tsconfig.json:
```diff
@@ -14,8 +14,8 @@
-    "strictNullChecks": true,
-    "noEmit": true
+    "noEmit": true,
+    "resolveJsonModule": true
```
This can be removed now
```suggestion
```

internettrans commented on tsconfig.json:
```diff
@@ -14,8 +14,8 @@
       "es2016.array.include",
       "es2018"
     ],
-    "strictNullChecks": true,
```
this should not be removed.

internettrans commented on src/openmrs-backend-dependencies.ts:
```diff
@@ -0,0 +1,92 @@
+  const modulesWithWrongVersionInstalled = getMismatchedVersions(
+    requiredBackendModules
+  );
+  console.error(
```
👍 this is a good start. In a future PR, let's surface these errors into a new tab of the devtools. Right now devtools only has the "Import Map" tab. Let's add a "Backend modules" tab in the future

<img width="697" alt="image" src="https://user-images.githubusercontent.com/5524384/65981568-9f179300-e436-11e9-88e5-73450a2a3f60.png">

internettrans commented on src/openmrs-backend-dependencies.ts:
```diff
@@ -0,0 +1,92 @@
+    missingOpenmrsBackendModules
+  );
+  console.error(
+    "The following backend modules versions are required",
```
```suggestion
    "The following backend module versions are required",
```

internettrans commented on src/openmrs-backend-dependencies.ts:
```diff
@@ -0,0 +1,92 @@
+    "The following backend modules are required!",
+    missingOpenmrsBackendModules
+  );
+  console.error(
```
Should we only console.log if the `modulesWithWrongVersionInstalled` array has a length > 0? Same goes for the `missingOpenmrsBackendModules` console error

internettrans reviewed (changes requested):
Nice work! This is close to being mergeable.

Have you verified that it's working with one of the other esm libraries? If so, that's awesome.

fatmali commented:
Hi @joeldenning , I added the changes you suggested. The reason the build is failing is because of the `strictNullChecks` in `ts-config.json` for this particular [line](https://github.com/openmrs/openmrs-esm-api/blob/master/src/shared-api-objects/user-has-access-react.component.tsx#L5) is `esm-api` . I will submit a fix for this line in another PR.

fatmali commented:
@joeldenning  Also, I confirmed this works for all esm modules except `@openmrs/esm-api` which doesn't appear on the SystemJS `onload` hook. Do you have any ideas why all other modules once loaded trigger the hook except `@openmrs/esm-api`?

fatmali committed:
Joel's review

fatmali commented on package.json:
```diff
@@ -33,6 +33,7 @@
     "@babel/preset-env": "^7.5.5",
     "@babel/preset-react": "^7.0.0",
     "@babel/preset-typescript": "^7.3.3",
+    "@openmrs/esm-api": "1.0.3",
```
`@openmrs/esm-api` 1.0.4 is broken, will work on figuring it out today

internettrans replied:
How is it broken? Typescript?

fatmali replied:
`openmrs-esm-api` can't find fhir.js, this is fixed by adding fhir.js in the dependencies for `openmrs-esm-api`

internettrans commented on src/openmrs-backend-dependencies.ts:
```diff
@@ -0,0 +1,100 @@
+
+function checkBackendDeps(module: any) {
+  if (isEmpty(installedBackendModules)) {
+    fetchInstalledBackendModules().then(({ data }) => {
```
We should handle the error here. Otherwise, we'll still have an uncaught promise exception. See https://github.com/xjamundx/eslint-plugin-promise/blob/master/docs/rules/catch-or-return.md for an explanation on why this is important.

Maybe add `.catch(err => {setTimeout(() => {throw err})})` for now. The esm-error-handling repo will surface a function in the future to better handle this.

fatmali replied:
good catch, thanks I'll do that.

internettrans reviewed (approved):
I requested one small change. Feel free to merge after you make that change.

internettrans commented:
Build failed - Please fix https://travis-ci.com/openmrs/openmrs-esm-devtools/builds/129981682

fatmali commented:
The reason this build is failing is mainly because of `esm-api`, check this comment for more info https://github.com/openmrs/openmrs-esm-devtools/pull/10#issuecomment-537353047

internettrans commented:
Why is the build failing if it's esm-api@1.0.4 that causes the problem but we're using esm-api@1.0.3??

fatmali committed:
Review changes

fatmali commented:
**updated PR with the following changes:**
1) removed strictNull checks from tsconfig - its the reason the build is failing due [this](https://github.com/openmrs/openmrs-esm-api/pull/14/files#r330676160) typescript issue in `openmrs-esm-api`

2) added a catch statement for all promises.
